script/tool/lib/src/pubspec_check_command.dart

```diff
@@ -243,6 +252,49 @@ class PubspecCheckCommand extends PackageLoopingCommand {
     return null;
   }
 
+  // Validates any "default_package" entries a plugin, returning an error
+  // string if there are any issues.
+  //
+  // Should only be called on plugin packages.
+  String? _checkForDefaultPackageError(
+    Pubspec pubspec, {
+    required RepositoryPackage package,
+  }) {
+    final dynamic platformsEntry = pubspec.flutter!['plugin']!['platforms'];
+    if (platformsEntry == null) {
+      logWarning('Does not implement any platforms');
+      return null;
+    }
+    final YamlMap platforms = platformsEntry as YamlMap;
+    final String packageName = package.directory.basename;
+
+    // Validate that the default_package entries look correct (e.g., no typos).
+    final Set<String> defaultPackages = <String>{};
+    for (final MapEntry<dynamic, dynamic> platformEntry in platforms.entries) {
+      final String? defaultPackage =
+          platformEntry.value['default_package'] as String?;
+      if (defaultPackage != null) {
+        defaultPackages.add(defaultPackage);
+        if (!defaultPackage.startsWith('${packageName}_')) {
+          return '"$defaultPackage" is not an expected implementation name '
+              'for "$packageName"';
+        }
```

Comment:
> Instead of this, why not check that defaultPackage is contained as a dependency of the plugin?

That's the very next block of code :)

> That way we could still endorse packages that have a name that doesn't start with packageName (because it might be taken, or because somebody else started the plugin before this rule went into effect?)

The reason I did this in addition is that I think it's useful to enforce prefix naming; even if the name we might normally use is taken, I think we should avoid the collision via suffix changes rather than breaking our core naming pattern.

Reply:
> That's the very next block of code :)

🤦 Gonna grab some coffee, my attention span is shrinking dangerously!

> even if the name we might normally use is taken, I think we should avoid the collision via suffix changes rather than breaking our core naming pattern.

Yeah, we've had to do it with some "_for_web" packages. I was thinking more on the case that somebody else has a plugin for some platform and they called it: "contoso_better_url_launcher_for_windows" or similar. That'll probably never happen, though :P